iceoryx_dds/test/moduletests/test_fastdds.cpp

```diff
@@ -0,0 +1,193 @@
+    EXPECT_EQ(true, takeResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::INVALID_RECV_BUFFER, takeResult.get_error());
+
+    auto takeNextResult = reader.takeNext(buffer, bufferSize);
+    EXPECT_EQ(true, takeNextResult.has_error());
+    EXPECT_EQ(iox::dds::DataReaderError::INVALID_RECV_BUFFER, takeNextResult.get_error());
+}
+
+TEST_F(FastDDSTest, SendAndTakeNext)
+{
+    uint64_t bufferSize = 1024;
+
+    /* Create reader */
+    TestDataReader reader{"test", "test", "test"};
+    reader.connect();
+
+    /* Create writer */
+    TestDataWriter writer{"test", "test", "test"};
+    writer.connect();
+
+    /* Wait for discovery */
+    writer.waitForReaderDiscovery(1);
+    reader.waitForWriterDiscovery(1);
+
+    /* Setup send buffer */
+    uint8_t sendBuffer[bufferSize];
+    // Populate send buffer with random values
+    std::srand(static_cast<uint>(std::time(nullptr)));
+    for (uint64_t i = 0; i < bufferSize; i++)
+    {
+        sendBuffer[i] = static_cast<uint8_t>(std::rand() % 255);
+    }
+
+    /* Send data */
+    writer.write(sendBuffer, bufferSize);
+
+    /* Wait for delivery */
+    reader.waitForData();
+
+    /* Setup receive buffer */
+    uint8_t receiveBuffer[bufferSize];
+    for (uint64_t i = 0; i < bufferSize; i++)
+    {
+        receiveBuffer[i] = 0;
+    }
+
+    /* Take data */
+    auto result = reader.takeNext(receiveBuffer, bufferSize);
+
+    /* Check correctness */
+    EXPECT_EQ(false, result.has_error());
+    for (uint64_t i = 0; i < bufferSize; i++)
+    {
+        ASSERT_EQ(sendBuffer[i], receiveBuffer[i]);
+    }
+}
+
+TEST_F(FastDDSTest, SendAndTake)
+{
+    uint64_t bufferSize = 1024;
+
+    /* Create reader */
+    TestDataReader reader{"test", "test", "test"};
+    reader.connect();
+
+    /* Create writer */
+    TestDataWriter writer{"test", "test", "test"};
+    writer.connect();
+
+    /* Wait for discovery */
+    writer.waitForReaderDiscovery(1);
+    reader.waitForWriterDiscovery(1);
+
+    /* Setup send buffers */
+    uint8_t sendBuffer1[bufferSize];
+    // Populate send buffer with random values
+    std::srand(static_cast<uint>(std::time(nullptr)));
+    for (uint64_t i = 0; i < bufferSize; i++)
+    {
+        sendBuffer1[i] = static_cast<uint8_t>(std::rand() % 255);
+    }
+
+    uint8_t sendBuffer2[bufferSize];
+    for (uint64_t i = 0; i < bufferSize; i++)
+    {
+        sendBuffer2[i] = static_cast<uint8_t>(std::rand() % 255);
+    }
+
+    /* Send data */
+    writer.write(sendBuffer1, bufferSize);
+    writer.write(sendBuffer2, bufferSize);
+
+    /* Wait for delivery */
+    reader.waitForData();
+
+    /* Setup receive buffer */
+    uint8_t receiveBuffer[bufferSize*2];
+    for (uint64_t i = 0; i < bufferSize*2; i++)
+    {
+        receiveBuffer[i] = 0;
+    }
+
+    /* Take data */
+    auto result = reader.take(receiveBuffer, bufferSize*2, 2);
+
+    /* Check correctness */
+    EXPECT_EQ(false, result.has_error());
```

Comment:
I think you should write `ASSERT_EQ` here  otherwise the test will run into undefined behavior realm when the result has an error. 

The other `ASSERT_EQ` you could replace with `EXPECT_EQ`.